source/common/network/connection_impl.cc

```diff
@@ -275,6 +275,13 @@ void ConnectionImpl::raiseEvent(ConnectionEvent event) {
     // connected events.
     callback->onEvent(event);
   }
+  // We may have pending data in the write buffer on transport handshake
+  // completion, which may also have completed in the context of onReadReady(),
+  // where no check of the write buffer is made. Provide an opportunity to flush
+  // here. If connection write is not ready, this is harmless.
+  if (event == ConnectionEvent::Connected && write_buffer_->length() > 0) {
```

Comment:
Should this happen before notifying callbacks? 

If we leave this after the above loop, should check whether the connection was closed before doing this.